ENH: optimisation of array_equal

Carreau opened the pull request:
For various reason I was looking at array_equal and I believe there are
a few case where we can get some speedup/less memory allocation.

  - if `equal_nan == True` then could we add a shortcut, we can sometime
    check that both array are the same and avoid the computation
  - if the dtypes cannot contain `nan`, then we do not need the expensive
    fancy indexing and removal of `nan`s

we can test will the following script

```python
import numpy as np

base = np.random.randint(1, 100, (10000, 10000), dtype='int64')
a = base.copy()
a1 = base.copy()

b = base.copy().astype('float')
b[b == 50] = np.nan

b1 = b.copy()

data = {'a':a, 'b':b, 'a1':a1, 'b1': b1}

for (x, y) in [('a','a'),('a','a1'),('b','b'),('b','b1')]:
    ax = data[x]
    bx = data[y]
    for en in [True, False]:
        res = np.array_equal(ax, bx, equal_nan=en)
        print(f"np.array_equal({x}, {y}, equal_nan={en}) == {res}")
        %timeit -r10 -o np.array_equal(ax, bx, equal_nan=en)

```

Which give us the following results.

```diff
--- before.txt  2023-09-07 15:19:01
+++ after.txt   2023-09-07 15:24:30
@@ -1,16 +1,16 @@
np.array_equal(a, a, equal_nan=True) == True
-419 ms ± 42.4 ms per loop (mean ± std. dev. of 10 runs, 1 loop each)
+1.41 µs ± 5.37 ns per loop (mean ± std. dev. of 10 runs, 1,000,000 loops each)

np.array_equal(a, a, equal_nan=False) == True
-31.1 ms ± 200 µs per loop (mean ± std. dev. of 10 runs, 10 loops each)
+31 ms ± 348 µs per loop (mean ± std. dev. of 10 runs, 10 loops each)

np.array_equal(a, a1, equal_nan=True) == True
-1.07 s ± 326 ms per loop (mean ± std. dev. of 10 runs, 1 loop each)
+40.3 ms ± 133 µs per loop (mean ± std. dev. of 10 runs, 10 loops each)

np.array_equal(a, a1, equal_nan=False) == True
-41.9 ms ± 130 µs per loop (mean ± std. dev. of 10 runs, 10 loops each)
+40.3 ms ± 210 µs per loop (mean ± std. dev. of 10 runs, 10 loops each)

p.array_equal(b, b, equal_nan=True) == True
-569 ms ± 61.7 ms per loop (mean ± std. dev. of 10 runs, 1 loop each)
+909 ns ± 1.61 ns per loop (mean ± std. dev. of 10 runs, 1,000,000 loops each)

np.array_equal(b, b, equal_nan=False) == False
-31.2 ms ± 208 µs per loop (mean ± std. dev. of 10 runs, 10 loops each)
+30.2 ms ± 228 µs per loop (mean ± std. dev. of 10 runs, 10 loops each)

np.array_equal(b, b1, equal_nan=True) == True
-842 ms ± 171 ms per loop (mean ± std. dev. of 10 runs, 1 loop each)
+461 ms ± 3.07 ms per loop (mean ± std. dev. of 10 runs, 1 loop each)

np.array_equal(b, b1, equal_nan=False) == False
-40.4 ms ± 383 µs per loop (mean ± std. dev. of 10 runs, 10 loops each)
+39.2 ms ± 234 µs per loop (mean ± std. dev. of 10 runs, 10 loops each)
```

seberg commented:
Just my first thoughts:  I am not sure I think it needs to be our job to check for identity, even when we know its correct.  I am happy to skip NaN checks when we know for sure that there are no NaNs.
One thing to beware of: timedeltas have NaNs, and I am not quite sure you are catching them (which I agree is wrong by NumPy if the case, but we need to take care here either way).

Carreau commented:
> I am not sure I think it needs to be our job to check for identity, even when we know its correct

Yeah, I'm unsure, I think it's pretty rare, but at the same time, it can be quite a bit more performant.


> One thing to beware of: timedeltas have NaNs, and I am not quite sure you are catching them (which I agree is wrong by NumPy if the case, but we need to take care here either way).

Do you mean we should have an explicit list of non-nanable types instead of issubtype ? 

Regardless both are not a hill I'm going to die on, I'm happy to split the logic in two PRs if you wish.

seberg commented:
I am just not sure that your current `issubdtype` (which I don't like, but there isn't a much cleaner alternative probably) doesn't also match for timedeltas: this needs a test.

seberg commented:
Actually an alternative that we may or may not like:  I always think that it might be useful to add a ufunc for [`islessgreater`](https://en.cppreference.com/w/c/numeric/math/islessgreater) or maybe an inverted version.  Maybe with a better name `not_different`?!

I.e. so that we don't have to do the NaN dance at all, but have a ufunc that matches NaNs (I think this would also help fix a bunch of `np.unique` annoyances/bugs).

Carreau committed:
ENH: optimisation of array_equal

For various reason I was looking at array_equal and I believe there are
a few case where we can get some speedup/less memory allocation.

 - if `equal_nan == True` then could we add a shortcut, we can sometime
   check that both array are the same and avoid the computation.
 - if the dtypes cannot contain `nan`, then we do not need the expensive
   fancy indexing and removal of `nan`s

we can test will the following script

```
import numpy as np

base = np.random.randint(1, 100, (10000, 10000), dtype='int64')
a = base.copy()
a1 = base.copy()

b = base.copy().astype('float')
b[b == 50] = np.nan

b1 = b.copy()

data = {'a':a, 'b':b, 'a1':a1, 'b1': b1}

for (x, y) in [('a','a'),('a','a1'),('b','b'),('b','b1')]:
    ax = data[x]
    bx = data[y]
    for en in [True, False]:
        res = np.array_equal(ax, bx, equal_nan=en)
        print(f"np.array_equal({x}, {y}, equal_nan={en}) == {res}")
        %timeit -r10 -o np.array_equal(ax, bx, equal_nan=en)

```

Which give us the following results.

```
--- before.txt	2023-09-07 15:19:01
+++ after.txt	2023-09-07 15:24:30
@@ -1,16 +1,16 @@
 np.array_equal(a, a, equal_nan=True) == True
-419 ms ± 42.4 ms per loop (mean ± std. dev. of 10 runs, 1 loop each)
+1.41 µs ± 5.37 ns per loop (mean ± std. dev. of 10 runs, 1,000,000 loops each)
 np.array_equal(a, a, equal_nan=False) == True
-31.1 ms ± 200 µs per loop (mean ± std. dev. of 10 runs, 10 loops each)
+31 ms ± 348 µs per loop (mean ± std. dev. of 10 runs, 10 loops each)
 np.array_equal(a, a1, equal_nan=True) == True
-1.07 s ± 326 ms per loop (mean ± std. dev. of 10 runs, 1 loop each)
+40.3 ms ± 133 µs per loop (mean ± std. dev. of 10 runs, 10 loops each)
 np.array_equal(a, a1, equal_nan=False) == True
-41.9 ms ± 130 µs per loop (mean ± std. dev. of 10 runs, 10 loops each)
+40.3 ms ± 210 µs per loop (mean ± std. dev. of 10 runs, 10 loops each)
 np.array_equal(b, b, equal_nan=True) == True
-569 ms ± 61.7 ms per loop (mean ± std. dev. of 10 runs, 1 loop each)
+909 ns ± 1.61 ns per loop (mean ± std. dev. of 10 runs, 1,000,000 loops each)
 np.array_equal(b, b, equal_nan=False) == False
-31.2 ms ± 208 µs per loop (mean ± std. dev. of 10 runs, 10 loops each)
+30.2 ms ± 228 µs per loop (mean ± std. dev. of 10 runs, 10 loops each)
 np.array_equal(b, b1, equal_nan=True) == True
-842 ms ± 171 ms per loop (mean ± std. dev. of 10 runs, 1 loop each)
+461 ms ± 3.07 ms per loop (mean ± std. dev. of 10 runs, 1 loop each)
 np.array_equal(b, b1, equal_nan=False) == False
-40.4 ms ± 383 µs per loop (mean ± std. dev. of 10 runs, 10 loops each)
+39.2 ms ± 234 µs per loop (mean ± std. dev. of 10 runs, 10 loops each)
```

Carreau commented:
The approach `issubdtype` does indeed fail for timestamp64, I decided to fall back to a `dtype in {...}` with an explicit list of dtype that cannot have NaNs, in a utility function that would make it easy to update this list. 

I've refactored the test to use pytest parametrize, and added a number of case of interest, and rebased to put the tests commit first (modified test is passing before the PR).

eendebakpt commented on numpy/core/numeric.py:
```diff
@@ -2442,6 +2445,16 @@ def array_equal(a1, a2, equal_nan=False):
+    if cannot_have_nan:
+        if a1 is a2:
+            return True
+        return bool(asarray(a1 == a2).all())
```
```suggestion
        return bool((a1 == a2).all())
```
Since `a1` and `a2` have already been cast with `asarray`

Carreau replied:
I'm not confident in removing the asarray, as I'm not sure why it's here ([Chesterton's fence](https://en.wikipedia.org/wiki/G._K._Chesterton#Chesterton's_fence)):
 - it is present on line 2477 – before this patch, 
 - and on line 2461 – last return of this function also present before this patch. 


From what I can tell this was introduced in #3322 (2013), long after this function (2006) see [this comment](https://github.com/numpy/numpy/pull/3322#discussion_r4183721).

>  If the two operands are 0d arrays, the == operator will return a bool object instead of an array, in which case the call to all() will fail.

Carreau replied:
I think this is still ok as now, 0d arrays comparison return np.True_ / np.False, but then if I remove I might as well remove also in the existing branches right ?

seberg replied:
It may be required for object array support, although ideall there is a test for that that fails if you remove it.

Carreau replied:
Ok, I pushed a commit to try.

Carreau commented:
Added a test for 0-d arrays in the first commit. 

I can remove the `asarray(? == ?)` in the function but I'm not entirely confident about it.

Carreau commented:
> .e. so that we don't have to do the NaN dance at all, but have a ufunc that matches NaNs (I think this would also help fix a bunch of `np.unique` annoyances/bugs

Maybe stupid question, I'ts not my area of expertice but may assuming array have same dtypes, woudln't `a.tobytes() == b.tobytes()` work for this ? I guess `tobytes()` make copies, but maybe there is a way to do it w/o copies ?

Carreau commented on numpy/core/tests/test_numeric.py:
```diff
@@ -1965,57 +1965,146 @@ def test_base_range(self):
+
+_md1 = np.array([[0, 1], [np.nan, 1]])
+
+# I belive this is a mistake, as both cplx ends up being nan+nan.j (same
+# complex in the list) twice instead of haveig nan+1.j, 1+nan.j
+_cplx1, _cplx2 = [np.array([1 + 1j])] * 2
+_cplx1.real, _cplx2.imag = np.nan, np.nan
+
+# this is I believe what was originally meant
+_cplx3, _cplx4 = np.complex64(1, np.nan), np.complex64(np.nan, 1)
```
I think @rossbar wrote the original test, maybe you remember what you meant ?

rossbar replied:
Ack yes, good catch. The intent was to assert that arrays are equal when either the real *or* the imaginary parts are nan, so the original test should be something like:

```python
>>> a = np.array([1 + 1j])
>>> b = a.copy()
>>> a.real = np.nan
>>> a
array([nan+1.j])
>>> b.image = np.nan
>>> b
array([1.+nanj])
>>> np.array_equal(a, b, equal_nan=True)
True
```

The [existing test](https://github.com/numpy/numpy/blame/5ca26b1b9df5e8c89923a2bb723a5ab95b6fe09b/numpy/core/tests/test_numeric.py#L2014-L2018) is correct, but indeed is not as specific as intended!

Carreau replied:
Great ! I'll update the message !

Carreau commented:
> Maybe stupid question, I'ts not my area of expertice but may assuming array have same dtypes, woudln't `a.tobytes() == b.tobytes()` work for this ? I guess `tobytes()` make copies, but maybe there is a way to do it w/o copies ?

Just tests, that does not work for `1+nan.j == nan + 1.j`, but I guess we have at least `a.tobytes() == b.tobytes()` =>  `a == b`. Maybe that's an extra optimisation.

ngoldbaum commented:
> woudln't a.tobytes() == b.tobytes() work for this 

I think it might work for all but object arrays since you can't know the truthiness of a python value without testing it.

Carreau commented:
Moving to ready for review. I'm happy to split into a multiple chunks, like only the tests for example.

eendebakpt commented on numpy/core/numeric.py:
```diff
@@ -2379,6 +2379,9 @@ def within_tol(x, y, atol, rtol):
 def _array_equal_dispatcher(a1, a2, equal_nan=None):
     return (a1, a2)
 
+def _dtype_cannot_hold_nan(dtype):
+    return dtype in {np.int8, np.int16, np.int32, np.int64}
```
```suggestion
_no_nan_dtypes = {np.int8, np.int16, np.int32, np.int64}

def _dtype_cannot_hold_nan(dtype):
    return dtype in _no_nan_dtypes
```
This is much faster, a benchmark:
```
import numpy as np

def _dtype_cannot_hold_nan(dtype):
    return dtype in {np.int8, np.int16, np.int32, np.int64}

_no_nan_dtypes = {np.int8, np.int16, np.int32, np.int64}

def _dtype_cannot_hold_nan_2(dtype):
    return dtype in _no_nan_dtypes

dtype=np.array([1.]).dtype

%timeit _dtype_cannot_hold_nan(dtype)
%timeit _dtype_cannot_hold_nan_2(dtype)
%timeit dtype in _no_nan_dtypes
```
results in
```
507 ns ± 63.1 ns per loop (mean ± std. dev. of 7 runs, 1,000,000 loops each)
112 ns ± 1.94 ns per loop (mean ± std. dev. of 7 runs, 10,000,000 loops each)
57.9 ns ± 5.47 ns per loop (mean ± std. dev. of 7 runs, 10,000,000 loops each)
```
In inline version is the fastest, so that would be an option as well.

Carreau replied:
One question is do we want to let people add things to `_no_nan_dtypes` by monkeypatching.

eendebakpt commented on numpy/core/numeric.py:
```diff
@@ -2379,6 +2379,9 @@ def within_tol(x, y, atol, rtol):
 def _array_equal_dispatcher(a1, a2, equal_nan=None):
     return (a1, a2)
 
+def _dtype_cannot_hold_nan(dtype):
+    return dtype in {np.int8, np.int16, np.int32, np.int64}
```
```suggestion
def _dtype_cannot_hold_nan(dtype):
    return dtype.type in {np.int8, np.int16, np.int32, np.int64}
```

Carreau replied:
Merging both your suggestions:

```suggestion
_no_nan_dtypes = {np.int8, np.int16, np.int32, np.int64}

def _dtype_cannot_hold_nan(dtype):
    return dtype.type in _no_nan_dtypes
```

eendebakpt commented on numpy/core/numeric.py:
```diff
@@ -2441,14 +2444,24 @@ def array_equal(a1, a2, equal_nan=False):
+    cannot_have_nan = (_dtype_cannot_hold_nan(a1.dtype) and
+                       _dtype_cannot_hold_nan(a2.dtype))
+    if cannot_have_nan:
+        if a1 is a2:
```
How common is the case of comparing identical (as in `id`) arrays? This slows down the other paths (although not a lot, the `a1 is a2` is quite fast)

Carreau replied:
As for many things... it depends on the usage. In napari for example quite a lot (but for small-ish array).
As the model it's event based when something change it regularly check the dimensions orders, masked data and bunch stuff like that to know wether it should redraw the screen.


The `a is b` case is actually the one the that drove me here in the first case.

Carreau commented:
For both the `is` question and the `dtype` micro-optimisation I think depends on wether you want to optimize for large or small arrays.

I'm tempted to want to optimize for large array (even if my use case is small-ish). This is were we'll see the most gain. Timing wise as we to at least µs, a few ns for dtype checking won't make much the differences. For large array checking, the `is` can have also a large impact on memory allocation.

eendebakpt commented on numpy/core/numeric.py:
```diff
@@ -2379,6 +2379,11 @@ def within_tol(x, y, atol, rtol):
 def _array_equal_dispatcher(a1, a2, equal_nan=None):
     return (a1, a2)
 
+def _dtype_cannot_hold_nan(dtype):
+    # can't extract set in module variable as it create a circular import.
+    # AttributeError: partially initialized module 'numpy' has no attribute 'int8' (most likely due to a circular import)
+    return dtype.type in {np.int8, np.int16, np.int32, np.int64}
```
```suggestion
_no_nan_types = None
def _dtype_cannot_hold_nan(dtype):
    global _no_nan_types
    if _no_nan_types is None:
        _no_nan_types = {np.bool_, np.int8, np.int16, np.int32, np.int64}
    return dtype.type in _no_nan_types
```
This avoids the circular import.

Carreau commented:
> There is still the question about the complex test that @rossbar wrote where I'm not sure what was written was what was intended.

Updated comments following discussion.

Carreau commented:
Going through my list of opened PRs, Just checking if there more I can do to help. 

I can pull the test refactor in a separate PRs to make this more digestible/less contentious.

Carreau commented:
Gentle nudge again, as this was relatively painful to rebase on master after the recent changes. 

I can again pull the test refactor into its own PR if that makes it easier to get accepted.

Carreau commented:
> It looks like some of the new tests are failing.

Yeah, I've botched the merge/rebase. Redoing now.

> If you'd like, stop by the next NumPy triage zoom meeting on Wednesday at 1:00 PM US Eastern time and you'll have more people's attention.

I'll try, but Those times are usually touch for me here, and today I'll can't for sure.

I've pulled out the test refactor into its own PR #25010

Carreau committed:
fix shadowing of None objects

seberg commented on numpy/_core/numeric.py:
```diff
@@ -2446,14 +2456,24 @@ def array_equal(a1, a2, equal_nan=False):
     if not equal_nan:
-        return bool(asarray(a1 == a2).all())
+        return bool((a1 == a2).all())
+    cannot_have_nan = _dtype_cannot_hold_nan(a1.dtype)\
+            and _dtype_cannot_hold_nan(a2.dtype)
```
```suggestion
    cannot_have_nan = (_dtype_cannot_hold_nan(a1.dtype)
                       and _dtype_cannot_hold_nan(a2.dtype)
```
Just avoid the `\` because I hate it :).

Carreau replied:
👍

seberg commented on numpy/_core/numeric.py:
```diff
@@ -2385,6 +2385,16 @@ def _array_equal_dispatcher(a1, a2, equal_nan=None):
+def _dtype_cannot_hold_nan(dtype):
+    global _no_nan_types
+    if _no_nan_types is None:
+        _no_nan_types = {np.bool_, np.int8, np.int16, np.int32, np.int64}
```
Could just use `import .numerictypes as nt`.  The problem is that things are wonky while we are setting things up.
I might slightly prefer (but verbose) to write:
`_no_nan_dtypes = {np.dtypes.BoolDType, np.dtypes.Int8DType, ...}`
and then compare `type(dtype)` rather than `dtype.type`.  We could put the `_no_nan_dtypes` into `np.dtypes` also.

I do *not* want to consider this public, though.  Because the right API should be different (maybe a `np.dtypes.NoNaNDType.register(MyDType)`,

Carreau replied:
Yeah, I agree that the API should not be public, at least for now. And, `numerictypes` seem great !

seberg commented on numpy/_core/numeric.py:
```diff
@@ -2505,7 +2525,7 @@ def array_equiv(a1, a2):
         return False
 
-    return bool(asarray(a1 == a2).all())
+    return bool((a1 == a2).all())
```
Nice! this does seem OK to remove, because we guarantee that `a == b` returns a NumPy boolean array/scalar.  So we don't run into problems even for object dtype!

Carreau replied:
I can also send that as it's own PR if you wish.

seberg commented on numpy/core/numeric.py:
```diff
@@ -1,8 +1,10 @@
-    ret = getattr(numeric, attr_name, None)
-    if ret is None and attr_name != "newaxis":
+
+    sentinel = object()
```
Not really this PR, but I don't think I care too much.

Carreau replied:
If I don't do this, test fails because `_no_nan_dtypes` is `None`.

seberg reviewed (approved):
I don't love these special cases, but I think they are fine.

Just nitpicks, I don't care what you do with them, although I think I would prefer to use the `nt.<type>` approach at least, because the global is just awkward.

Carreau committed:
review suggestions

Carreau commented:
Here are some redone bench test with 1000x1000 random array: 
![BENCH](https://github.com/numpy/numpy/assets/335567/aef4e01d-3154-466f-a33d-c13cfbc0d8c7)

It's only a guideline as the larger the arrays are the more the effect will be felt.

(Source to replicate : https://gist.github.com/Carreau/baecfc0b02773bea6a744935e0dc9d1b)

CI seem to have a problem with np.dtypes, but it works locally, which is weird.

Carreau committed:
try numerictype types

seberg commented on numpy/_core/numeric.py:
```diff
@@ -2385,6 +2385,19 @@ def _array_equal_dispatcher(a1, a2, equal_nan=None):
+    np.dtypes.Int16DType,
+    np.dtypes.Int32DType,
+    np.dtypes.Int64DType,
+}
```
Strange that I cannot repro locally, but the try to import modules lazily seems to make things barf.  Importing `dtypes` lazily is useless (The C side has to already import it to fill it!), but I guess explicitly importing here will work...

seberg replied:
It is actually a bit annoying, because `np.longlong` or `np.cint` could fall through the cracks (right now), but will ignore that.

Carreau committed:
try reimport

Carreau commented:
I'm not quite sure why arm64 is failing, can you restart it ? Or I can merge main/rebase

seberg commented on numpy/_core/numeric.py:
```diff
@@ -7,7 +7,9 @@
 import builtins
 
 import numpy as np
+import numpy.dtypes
```
```suggestion
```

seberg commented on numpy/_core/numeric.py:
```diff
@@ -2385,6 +2387,19 @@ def _array_equal_dispatcher(a1, a2, equal_nan=None):
 
 
+_no_nan_types = {
+    nt.bool,
+    nt.int8,
+    nt.int16,
+    nt.int32,
+    nt.int64,
```
```suggestion
    # should use np.dtype.BoolDType, but as of writing
    # that fails the reloading test.
    type(np.dtype(nt.bool)),
    type(np.dtype(nt.int8)),
    type(np.dtype(nt.int16)),
    type(np.dtype(nt.int32)),
    type(np.dtype(nt.int64)),
```

Carreau replied:
that fails as well with `numpy has not attribute dtype`

seberg replied:
bleh, yah... but it seems explicitly imported already, so no problem...

seberg commented on numpy/_core/numeric.py:
```diff
@@ -2385,6 +2387,19 @@ def _array_equal_dispatcher(a1, a2, equal_nan=None):
+
+
+def _dtype_cannot_hold_nan(dtype):
+    return dtype in _no_nan_types
```
```suggestion
    return type(dtype) in _no_nan_types
```

seberg reviewed (commented):
Sorry forgot about this a bit.  The `np.dtypes` namespace population hates reloading, so let's use the ugly way to get the same thing.  The `in` check needs to work on types, so either `dtype.type` or `type(dtype)` (depending on what is used).

seberg committed:
Apply suggestions from code review

seberg commented on numpy/_core/numeric.py:
```diff
@@ -2385,6 +2386,21 @@ def _array_equal_dispatcher(a1, a2, equal_nan=None):
+_no_nan_types = {
+    # should use np.dtype.BoolDType, but as of writing
+    # that fails the reloading test.
+    type(np.dtype(nt.bool)),
+    type(np.dtype(nt.int8)),
+    type(np.dtype(nt.int16)),
+    type(np.dtype(nt.int32)),
+    type(np.dtype(nt.int64)),
```
```suggestion
    type(dtype(nt.bool)),
    type(dtype(nt.int8)),
    type(dtype(nt.int16)),
    type(dtype(nt.int32)),
    type(dtype(nt.int64)),
```

seberg committed:
Update numpy/_core/numeric.py

seberg commented:
Thanks @Carreau passing now, and my last set of changes were pretty small.

Carreau commented:
🥳 🎊  Thanks !

seberg commented:
Just to note in case anyone has time.  We need to tweak this a bit and add a regression test.  This unfortunately broke some comparisons of scalar arrays with `array_equal`.

(I think the reason is only historic special paths that lead to real booleans rather than NumPy ones... but what can you do...)

See gh-27271

eendebakpt commented:
@seberg Adding back the `asarray` would probably fix this. I assume you do not want to attempt the other option: update the code so that `np.array("foo") == np.array(1)` returns a numpy bool?

seberg commented:
Best not for a backport maybe. But happy for main branch.

eendebakpt commented:
I created https://github.com/numpy/numpy/pull/27275
